cpp/src/arrow/util/thread-pool.h

```diff
@@ -0,0 +1,147 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef ARROW_UTIL_THREAD_POOL_H
+#define ARROW_UTIL_THREAD_POOL_H
+
+#include <condition_variable>
+#include <deque>
+#include <exception>
+#include <functional>
+#include <future>
+#include <list>
+#include <memory>
+#include <mutex>
+#include <thread>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "arrow/status.h"
+#include "arrow/util/macros.h"
+
+namespace arrow {
+
+// Set the number of worker threads used by the process-global thread pool
+// for CPU-bound tasks.
+ARROW_EXPORT Status SetCPUThreadPoolCapacity(size_t threads);
+
+namespace internal {
+
+namespace detail {
+
+// Needed because std::packaged_task is not copyable and hence not convertible
+// to std::function.
+template <typename R, typename... Args>
+struct packaged_task_wrapper {
+  using PackagedTask = std::packaged_task<R(Args...)>;
+
+  explicit packaged_task_wrapper(PackagedTask&& task)
+      : task_(std::make_shared<PackagedTask>(std::forward<PackagedTask>(task))) {}
+
+  void operator()(Args&&... args) { return (*task_)(std::forward<Args>(args)...); }
+  std::shared_ptr<PackagedTask> task_;
+};
+
+}  // namespace detail
+
+class ThreadPool {
```

Comment:
Should this class appear in the Doxygen docs? If so, should use three forward slashes `///` and also add some tags like `\brief` to the function descriptions